components/brave_rewards/resources/page/reducers/promotion_reducer.ts

```diff
@@ -0,0 +1,234 @@
+        break
+      }
+
+      chrome.send('brave_rewards.attestPromotion', [promotionId, JSON.stringify({
+        captchaId: currentPromotion.captchaId,
+        x: parseInt(payload.x, 10),
+        y: parseInt(payload.y, 10)
+      })])
+      break
+    }
+    case types.RESET_PROMOTION: {
+      const promotionId = payload.promotionId
+      if (!state.promotions || !promotionId) {
+        break
+      }
+
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion) {
+        break
+      }
+
+      const promotions = state.promotions.map((item: Rewards.Promotion) => {
+        if (currentPromotion.promotionId === item.promotionId) {
+          item.captchaStatus = null
+          item.captchaImage = ''
+          item.captchaId = ''
+          item.hint = ''
+        }
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.DELETE_PROMOTION: {
+      const promotionId = payload.promotionId
+      if (!state.promotions || !promotionId) {
+        break
+      }
+      const currentPromotion = getPromotion(promotionId, state.promotions)
+
+      if (!currentPromotion) {
+        break
+      }
+
+      let promotions = state.promotions.map((item: Rewards.Promotion) => {
+        if (currentPromotion.promotionId === item.promotionId) {
+          item.captchaStatus = null
+          item.status = 4
+        }
+
+        return item
+      })
+
+      state = {
+        ...state,
+        promotions
+      }
+      break
+    }
+    case types.ON_PROMOTION_FINISH: {
+      state = { ...state }
+      let newPromotion: any = {}
+      const result = payload.properties.result
+
+      if (!state.promotions) {
+        break
+      }
+
+      let promotionId = payload.properties.promotion.promotionId
+      newPromotion.promotionId = promotionId
+
+      switch (result) {
+        case 0:
+          let ui = state.ui
+          newPromotion.captchaStatus = 'finished'
+          newPromotion.captchaImage = ''
+          newPromotion.captchaId = ''
+          newPromotion.hint = ''
+          ui.emptyWallet = false
+
+          state = {
+            ...state,
+            ui
+          }
+
+          chrome.send('brave_rewards.getWalletProperties')
+          chrome.send('brave_rewards.fetchBalance')
+          break
+        case 6:
+          newPromotion.captchaStatus = 'wrongPosition'
+          if (!promotionId) {
+            break
+          }
+          chrome.send('brave_rewards.claimPromotion', [promotionId])
+          break
+        default:
+          newPromotion.captchaStatus = 'generalError'
+          break
+      }
+
+      if (state.promotions) {
+        const promotions = updatePromotions(newPromotion, state.promotions)
+
+        state = {
+          ...state,
+          promotions
+        }
+      }
+
+      break
+    }
+  }
+  return state
+}
```

Comment:
Again, would recommend breaking up into functions per case

Reply:
right now we do this is in all reducers. Can you log an issue and can we can refactor all places